src/TuxxState.js

```diff
@@ -0,0 +1,340 @@
+  var newState = {};
+
+  // recurseKeys FUNCTION: traverses through keys in a heavily nested object, building our newState object as we iterate through potential nested objects and setting the deepest keys passed by newProps to be the result of invoking our callback on the deepest keys in the newProps and corresponding currentState keys
+  // @param currentST OBJECT: properties of the currentState
+  // @param newST OBJECT: properties of the newState
+  // @param newPR OBJECT: properties of the newProps
+  var recurseKeys = function (currentSt, newSt, newPr) {
+    for (var key in newPr) {
+      if (newPr.hasOwnProperty(key)) {
+        var valAtKey = newPr[key];
+        var stAtKey = currentSt[key];
+        // if the value at this newProps key is defined and is an object, and if it matches the corresponding key in the current state which is also defined and is an object
+        if (valAtKey !== null && typeof valAtKey === 'object' && !Array.isArray(valAtKey) && stAtKey !== null && typeof stAtKey === 'object' && !Array.isArray(stAtKey)) {
+          // shallow copy the keys from the currentState into our newState so that there are no references to the currentState keys, but our newState contains identical values
+          newSt[key] = assign({}, stAtKey);
+          // invoke recurseKeys with the nested objects inside the current key
+          recurseKeys(stAtKey, newSt[key], valAtKey);
+        } else {
+          // we have found a deep key in newProps and all keys match so far, so we invoke our callback, which sets the result of invoking our callback with these deep keys to be the value at the corresponding key in our newState
+          callback(newSt, key, stAtKey, valAtKey);
+        }
+      }
+    }
+  };
+  // invoke recurseKeys and build newState object by passing in the currentState, our empty newState, and newProps
+  recurseKeys(currentState, newState, newProps);
+  return newState;
+};
+
+// invariantNumberCheck FUNCTION: internal function that ensures a given value is a number or throws an error
+// @param input UNKNOWN: value we are performing invariant check upon
+var invariantNumberCheck = function (input) {
+  if (typeof(input) !== "number") {
+    invariant(!input, 'Cannot perform operation on "%s" because it is not of type number.', input);
+  };
+};
+
+// invariantArrayCheck FUNCTION: internal function that ensures a given value is an array or throws an error
+// @param input UNKNOWN: value we are performing invariant check upon
+var invariantArrayCheck = function (input) {
+  if (!Array.isArray(input)) {
+    invariant(!input, 'Cannot perform operation on "%s" because it is not an array.', input);
+  };
+};
+
+// invariantValueCheck FUNCTION: internal function that ensures a given value is not stored in an array or an object
+// @param input UNKNOWN: value we are performing invariant check upon
+var invariantValueCheck = function (input) {
```

Comment:
Recommend changing this to (the first argument to invariant might need to be inverted):

var invariantValueCheck = function (input) {
  var isObject = typeof(input) === "object";
  invariant(isObject, 'Cannot perform operation on "%s" because it must not be an array or object.', input);
};

Apply to all invariant checks.